### ISSUE TITLE
[HOLD for payment 2024-10-25] [HOLD for payment 2024-10-24] [$250] Android - Room  -On entering multiline text with header markdown, 2nd line is not visible

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#48281]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.26
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4904629
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Launch app
2. Tap on a room chat
3. Tap header -- description
4. Enter a multiline text with header markdown

## Expected Result:
On entering multiline text with header markdown, 2nd line must be visible.

## Actual Result:
On entering multiline text with header markdown, 2nd line is not visible.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/8799e4d1-a06a-4eec-a6f9-74f06bb75f3a

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0166150593d62a36c2</li>
        <li>Upwork Job ID: 1829289632232865933</li>
        <li>Last Price Increase: 2024-09-19</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 104139938</li>
<li>tsa321 | Contributor | 104139940</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by IuliiaHerets** [comment]
@trjExpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by trjExpensify** [comment]
Reproducible: 
<img width="375" alt="image" src="https://github.com/user-attachments/assets/e1e6701f-49f2-49b3-a938-5d7954b18603">

**Comment by jp928** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The room description on Android is cut off when typein multiple lines.

### What is the root cause of that problem?

https://github.com/Expensify/App/blob/9e82ed89b07c62ccb5d21a70398975bf0c37d11b/src/styles/utils/index.ts#L1242

function `getAutoGrowHeightInputStyle` doesn't workout a correct height. Consider to use `maxHeight: '100%'` instead, because we have clamp height on the container element. 
https://github.com/Expensify/App/blob/9e82ed89b07c62ccb5d21a70398975bf0c37d11b/src/styles/index.ts#L1158



### What changes do you think we should make in order to solve the problem?
function `getAutoGrowHeightInputStyle` doesn't workout a correct height. Consider to use `maxHeight: '100%'` instead, because we have clamp height on the container element. 

### What alternative solutions did you explore? (Optional)
NA
**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Room -On entering multiline text with header markdown, 2nd line is not visible

### What is the root cause of that problem?
To fix scroll bar flickering issue on line breaks we are the text input height here
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/components/TextInput/BaseTextInput/index.tsx#L386-L389
So the input has a fixed height so to allow autoGrow of height for multiline we are setting the height of the container to `textInputHeight` here
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/components/TextInput/BaseTextInput/index.tsx#L300
this `textInputHeight` is determined by the layout of an invisible `Text` here
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/components/TextInput/BaseTextInput/index.tsx#L512
But in our case the `InputComponent` is `RNMarkdownTextInput` which styles the text value based on the markdown and in this case when the user inputs header markdown it is bold and takes more space/width so jumps to the next line with fewer characters than where the normal invisible `Text` component discussed above which we use to calculate the `textInputHeight`.
### What changes do you think we should make in order to solve the problem?

The scroll-bar flickering problem is only for normal `RNTextInput` and doesn't occur for `RNMarkdownTextInput` so we need only to apply manual height for `!isMarkdownEnabled` case
So change
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/styles/utils/index.ts#L1229-L1242
```
getAutoGrowHeightInputStyle: (textInputHeight: number, maxHeight: number, includeHeight = true): ViewStyle => {
        if (textInputHeight > maxHeight) {
            return {
                ...styles.pr0,
                ...styles.overflowAuto,
            };
        }

        return {
            ...styles.pr0,
            ...styles.overflowHidden,
            // maxHeight is not of the input only but the of the whole input container
            // which also includes the top padding and bottom border
            ...(includeHeight && {['height']: maxHeight - styles.textInputMultilineContainer.paddingTop - styles.textInputContainer.borderBottomWidth}),
        };
```
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/styles/index.ts#L1158-L1162
```
autoGrowHeightInputContainer: (textInputHeight: number, minHeight: number, maxHeight: number, includeHeight = true) =>
            ({
                ...(includeHeight && {height: lodashClamp(textInputHeight, minHeight, maxHeight)}),
                ...(!includeHeight && {maxHeight}),
                minHeight,
            } satisfies ViewStyle),
```
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/components/TextInput/BaseTextInput/index.tsx#L300
```
autoGrowHeight &&
                            styles.autoGrowHeightInputContainer(
                                textInputHeight,
                                variables.componentSizeLarge,
                                typeof maxAutoGrowHeight === 'number' ? maxAutoGrowHeight : 0,
                                !isMarkdownEnabled,
                            ),
```
https://github.com/Expensify/App/blob/e66b1238e7ad0e69b0a12f429188d71bef09e35e/src/components/TextInput/BaseTextInput/index.tsx#L388
```
 ? [
                                              StyleUtils.getAutoGrowHeightInputStyle(textInputHeight, typeof maxAutoGrowHeight === 'number' ? maxAutoGrowHeight : 0, !isMarkdownEnabled),
                                              styles.verticalAlignTop,
                                          ]
```
### What alternative solutions did you explore? (Optional)

**Comment by trjExpensify** [comment]
@alitoshmatov can you give these proposals a review, please?

**Comment by alitoshmatov** [comment]
@jp928 Thank you for your proposal, not sure what you mean exactly but maxHeight is used to calculate input height based on text, padding and other factors. Setting it to `100%` in `getAutoGrowHeightInputStyle` function is not possible since it is expecting number and doing calculations on it.

**Comment by alitoshmatov** [comment]
@FitseTLT Thank you for your proposal, your RCA is correct. But your solution is not working on native app.

https://github.com/user-attachments/assets/f5f33928-4811-427c-89cc-35f7431614ff

**Comment by alitoshmatov** [comment]
@FitseTLT I think I added all needed changes I even checkout out to your provided branch before applying changes. Maybe I missed something, can you make sure changes are working in you ios native app and if yes provide me with detailed change for native too or you can update your branch

**Comment by alitoshmatov** [comment]
@excile1 Thank you for your proposal, your RCA is correct. Your solution looks to be working fine, the 2nd bug you specified does not seem to be present in ios app. I just applied your first solution and ios app is working fine. Let me do some research for on the second bug and other platforms

**Comment by FitseTLT** [comment]
> @FitseTLT I think I added all needed changes I even checkout out to your provided branch before applying changes. Maybe I missed something, can you make sure changes are working in you ios native app and if yes provide me with detailed change for native too or you can update your branch


@alitoshmatov I didn't add the changes in native files in my branch, sorry that's why I included a note on my proposal to add it but now I have updated my test [branch](https://github.com/FitseTLT/App/tree/fix-markdown-textinput-cuttoff-bug) by include the same changes for native platforms 👍 U can check.

**Comment by alitoshmatov** [comment]
@FitseTLT Still not working, can you share a screen recording of it working on your ios app

**Comment by alitoshmatov** [comment]
@tsa321 Thank you for your proposal, Your solution is working on ios fine, but have some issues on android
https://github.com/user-attachments/assets/5c0f19f4-503b-466b-a885-78e447b412ac

**Comment by trjExpensify** [comment]
Sounds good! Don't hesitate to use Slack to collab with some peeps if you need to.

**Comment by trjExpensify** [comment]
Great, can you let us know when you test it? Thanks!

**Comment by trjExpensify** [comment]
@alitoshmatov let us know your thoughts on next steps. 👍

**Comment by alitoshmatov** [comment]
https://github.com/Expensify/App/issues/48281#issuecomment-2349541711

Looks like a good plan. Based on [this](https://github.com/Expensify/App/issues/48281#issuecomment-2351236819) we can continue with @excile1.

 Though I think we get third opinion for solution with `react-native-live-markdown`.

**Comment by alitoshmatov** [comment]
@tomekzaw Looks like you have worked on `react-native-live-markdown`'s part that concerns us and also maintaining this package. So, would be great if we could get your opinion on this one

TLDR: The problem is in android where you can see little jump when heading markdown is written in the input. 

https://github.com/user-attachments/assets/5c0f19f4-503b-466b-a885-78e447b412ac


The solution proposed is:
> Fixing the second problem will require a separate issue/PR in the @expensify/react-native-live-markdown repository. For this test, I commented out the problematic line, but the proper solution I'm thinking of is to either try to extrapolate the lineHeight based on how much the fontSize changed, but I think the best solution would be to remove the default behavior of forcing lineHeight to 1.5x, while still allowing to provide a custom lineHeight in the markdown style.

https://github.com/Expensify/react-native-live-markdown/blob/037a1b0107263314ad5cdc710fa862280f65ea32/android/src/main/java/com/expensify/livemarkdown/MarkdownUtils.java#L148-L157

**Comment by alitoshmatov** [comment]
@excile1  I just noticed in the video you provided, jittering effect is still there when the line is full and goes to the next line

**Comment by tomekzaw** [comment]
Hi, author of react-native-live-markdown here. We're aware of this issue and we're working on a fix.

This problem appears when both `react-native-live-markdown` and `react-native-reanimated` are used. That's because each of these libraries has its own commit hook. It looks like RNLiveMarkdown's commit hook is registered first and RNReanimated's commit hook is registered as a second. Note that this problem would most likely not happen if the order of commit hooks was different (however we cannot rely on that and the solution needs to be order-agnostic).

In RNLiveMarkdown there's a `MarkdownCommitHook` that calls `AndroidTextInputShadowNode::setTextLayoutManager` method to change the algorithm used for calculating `MarkdownTextInput`'s height so that it reflects Markdown formatting. Unfortunately, this is information is effectively lost because in next step, this `AndroidTextInputShadowNode` is cloned again by `ReanimatedCommitHook`. When shadow node is cloned, `AndroidTextInputComponentDescriptor::adopt` method is cloned which also calls `setTextLayoutManager` with the original text layout manager that doesn't take into account Markdown formatting.

At this point we're still not sure why `AndroidTextInputShadowNode` is cloned again as a result of `ReanimatedCommitHook` call even if there are no running animations. We're currently investigating that. I also confirmed that this is indeed the root cause of the issue because the height is correct when I comment out `ReanimatedCommitHook`.

**Comment by j-piasecki** [comment]
Hey! I was helping @tomekzaw investigate the issue and what he said in the above message turned out not to be entirely true. The conflict between commit hooks from `react-native-reanimated` and `react-native-live-markdown` isn't impacting the App (it's dependent on the ordering of imports, if Live Markdown is imported after Reanimated, the problem doesn't occur) but visually it produces really similar results which is what led us off.

As was already stated in multiple proposals above, the root of the issue here is `autoGrow` implementation which uses `Text` components to calculate the height of the `TextInput`. Since it doesn't use any markdown styling, it returns wrong values when markdown is used.

Regarding the issues from the video, the little jump when creating a heading:
|Android|iOS|
|-|-|
|<video src="https://github.com/user-attachments/assets/8f0e2e4e-aeac-4bf7-98bb-c9a25fd7e0d9">|<video src="https://github.com/user-attachments/assets/8afcb396-a081-4220-a593-2b16bd0a9060">|

is expected - until a letter is written after `# ` it's not considered a heading so typing anything here results in changing styles for that paragraph. Since the font size for headings is different from the normal text on Android it also changes the height. One thing that could be improved here is updating the parser so that a space after `#` is enough for it to be considered a heading.

The other jump

https://github.com/user-attachments/assets/7057aac9-41b0-4219-93f2-e1d4e2dbc48b

as @tomekzaw explained above is coming from React Native itself. It's also happening on pure `TextInput` components and is caused by the layout being calculated on a different thread than the text is being updated.

I'd argue that both of these are outside the scope of this issue.

**Comment by tomekzaw** [comment]
As for the other jump mentioned by @j-piasecki, here's a video recording that compares `MarkdownTextInput` and RN's `TextInput` side-by-side: https://github.com/Expensify/react-native-live-markdown/issues/36#issuecomment-2360869552

**Comment by alitoshmatov** [comment]
@tomekzaw @j-piasecki Thank you both for these helpful insights.

**Comment by alitoshmatov** [comment]
So I think we can conclude that this issue is outside of our scope and we should go on with our existing solution to fix height calculation problem. @tsa321 Since we are only using your solution, will you be able to work on this yourself?

**Comment by alitoshmatov** [comment]
We can go with @tsa321 's [proposal](https://github.com/Expensify/App/issues/48281#issuecomment-2323218965)

C+ reviewed 🎀 👀 🎀

The issue in android where input height changes and jittery jump is known `react-native-live-markdown` issues

**Comment by justinpersaud** [comment]
Sounds good to me!

**Comment by tomekzaw** [comment]
> as @tomekzaw explained above is coming from React Native itself. It's also happening on pure TextInput components and is caused by the layout being calculated on a different thread than the text is being updated.

Reported as https://github.com/facebook/react-native/issues/46813

**Comment by trjExpensify** [comment]
Payments are due tomorrow. Checklist time, please. Thanks!

**Comment by trjExpensify** [comment]
Bump on the checklist!

**Comment by alitoshmatov** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/39519
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/39519/files#r1818147542
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: No need
- Determine if we should create a regression test for this bug.

**Regression Test Proposal**
1. Go to any room
2. Go to room description
3. Enter multi-line title using markdown like `# Title`
4. Make sure input height is correctly adjusting to new lines, and lines is not cut in half

**Comment by trjExpensify** [comment]
Thanks! Confirming payments made as follows, accounting for a deduction for 1 regression:

- $125 to @tsa321 for the fix
- $125 to @alitoshmatov for the review

This was caught with an existing regression test linked in the OP, so I'm not adding a new one. Paid, and closing!
